coil-base/src/main/java/coil/network/NetworkObserver.kt

```diff
@@ -1,51 +1,146 @@
 import coil.util.Logger
+import coil.util.isPermissionGranted
 import coil.util.log
 
 /**
- * This class caches the network state from [NetworkObserverStrategy].
+ * Observes the device's network state and calls [Listener] if any state changes occur.
  *
- * Instances should be cleaned up with [shutdown].
+ * This class provides a raw stream of updates from the network APIs. The [Listener] can be
+ * called multiple times for the same network state.
  */
-internal class NetworkObserver(
-    context: Context,
-    private val logger: Logger?
-) : NetworkObserverStrategy.Listener {
+internal interface NetworkObserver {
 
     companion object {
         private const val TAG = "NetworkObserver"
-        private const val ONLINE = "ONLINE"
-        private const val OFFLINE = "OFFLINE"
+
+        /** Create a new [NetworkObserver] instance. */
+        operator fun invoke(context: Context, listener: Listener, logger: Logger?): NetworkObserver {
+            val connectivityManager: ConnectivityManager? = context.getSystemService()
+            if (connectivityManager == null || !context.isPermissionGranted(ACCESS_NETWORK_STATE)) {
+                logger?.log(TAG, Log.WARN) { "Unable to register network observer." }
+                return EmptyNetworkObserver
+            }
+
+            return try {
+                if (SDK_INT >= 21) {
+                    NetworkObserverApi21(connectivityManager, listener)
+                } else {
+                    NetworkObserverApi14(context, connectivityManager, listener)
+                }
+            } catch (e: Exception) {
+                logger?.log(TAG, RuntimeException("Failed to register network observer.", e))
+                EmptyNetworkObserver
+            }
+        }
     }
 
-    private val strategy = NetworkObserverStrategy(context, this, logger)
+    /** Synchronously checks if the device is online. */
+    val isOnline: Boolean
+
+    /** Stop observing network changes. */
+    fun shutdown()
 
-    private var isOnline = strategy.isOnline()
-    private var isShutdown = false
+    /** Calls [onConnectivityChange] when a connectivity change event occurs. */
+    interface Listener {
+
+        @MainThread
+        fun onConnectivityChange(isOnline: Boolean)
+    }
+}
+
+private object EmptyNetworkObserver : NetworkObserver {
+
+    override val isOnline = true
+
+    override fun shutdown() {}
+}
+
+@RequiresApi(21)
+@SuppressLint("MissingPermission")
+private class NetworkObserverApi21(
+    private val connectivityManager: ConnectivityManager,
+    private val listener: Listener
+) : NetworkObserver {
+
+    private val networkCallback = object : ConnectivityManager.NetworkCallback() {
+        override fun onAvailable(network: Network) = onConnectivityChange(network, true)
+        override fun onLost(network: Network) = onConnectivityChange(network, false)
+    }
+
+    override val isOnline: Boolean
+        get() = connectivityManager.allNetworks.any { it.isOnline() }
```

Comment:
Will this throw an exception if the permission isn't allowed, or is it safe?

Reply:
It should be safe since we ensure we have the permissions [here](https://github.com/coil-kt/coil/pull/385/files#diff-6a3f731cb10b6dd03ddc46d173999f1dR37).